## 1.1 Installing librat

This tutorial runs through the basics of running librat.  Material is taken from <a href=https://librat.wikispaces.com/>librat wikispace</a>
<p>Librat is usually run from the command line, but to make things clearer, easier to run and more interactive, this Jupyter work book uses Python to call librat using the subprocess function.

## Prerequisites

<p><b>Numpy</b> can be installed using PIP</p>
<p><b>Matplotlib</b> can be installed using PIP</p>
<p>These usually ship with installations such as anaconda

## Download librat and install
Installing librat can be done on unix-like operating system (tested on Solaris, various linux, Mac OS X, etc.). <b>Librat will not install on a Windows OS</b>. To install librat download the <a href=http://www2.geog.ucl.ac.uk/~plewis/librat/configure>configure script</a>.  Then change the permission to make it an executable, this is done on the command line:
<code><span style="background-color:lightgrey">
chmod +x configure
</span></code>

Install librat with:
<code><span style="background-color:lightgrey">
./configure
</span></code>
This may take a couple of minutes

## Update environment variables
There are also a few environment variables that need to be updated:
<code><span style="background-color:lightgrey">
export BPMS=${HOME}/bpms
export PATH=$BPMS/bin/x86_64:$BPMS/bin/csh:$PATH
</span></code>

For Mac users:
<code><span style="background-color:lightgrey">
export DYLD_LIBRARY_PATH=${BPMS}/src/lib
</span></code>

For Linux ueser:
<code><span style="background-color:lightgrey">
export LD_LIBRARY_PATH=${BPMS}/src/lib
</span></code>

These can be added to your <code><span style="background-color:lightgrey">.bash_profile</span></code> or <code><span style="background-color:lightgrey">.bashrc</span></code>.  For this iPython session we can add these below.

In [1]:
import platform
import os

os.environ['BPMS'] = os.path.join(os.path.expanduser("~"), 'bpms')
os.environ['PATH'] += ':' + os.path.join(os.environ['BPMS'], 'bin/x86_64')
os.environ['PATH'] += ':' + os.path.join(os.environ['BPMS'], 'bin/csh')

if platform.system() == 'Darwin': # A Mac:
    os.environ['DYLD_LIBRARY_PATH'] = os.path.join(os.environ['BPMS'], 'src/lib')
else:
    os.environ['LD_LIBRARY_PATH'] = os.path.join(os.environ['BPMS'], 'src/lib')

You could also <code>crtl+z</code> out of your iPython session, update your <code><span style="background-color:lightgrey">.bash_profile</span></code> or <code><span style="background-color:lightgrey">.bashrc</span></code> then <code>source</code> your <code><span style="background-color:lightgrey">.bash_profile</span></code> or <code><span style="background-color:lightgrey">.bashrc</span></code> and then <code>fg</code> back into iPython

## Run a test

In [2]:
# this is an inbulit simple test we can run...
import subprocess
os.chdir(os.path.join(os.environ['BPMS'], 'src/start/'))

stdout = subprocess.Popen('./testMe', stdout=subprocess.PIPE)
for line in stdout.stdout:
    print line.strip()

BPMS environment variable set to /Users/kimcalders/bpms
this result
RTD 0
order: 0	intersection point:	0.000000 0.000000 1.000010
ray length:		98.999990
intersection material:	6
sun 0:			1 reflectance
lengthToSun: 9.000000 angleToSun 0.000000
direct:			0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
sky  :			reflectance
diffuse:		0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
wavebands: 0.000000 1.000000 2.000000 3.000000 4.000000 5.000000 6.000000
x: -0.499900 0.500100
y: -19.996000 20.004000
z: -0.999800 1.000200
bbox centre @ 0.000100 0.004000 0.000200
diff of file:


If the output above looks like this:

<code>
this result
<code>===========</code>
RTD 0
order: 0	intersection point:	0.000000 0.000000 1.000010
ray length:		98.999990
intersection material:	6
sun 0:			1 reflectance
lengthToSun: 9.000000 angleToSun 0.000000
direct:			0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
sky  :			reflectance
diffuse:		0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
wavebands: 0.000000 1.000000 2.000000 3.000000 4.000000 5.000000 6.000000
x: -0.499900 0.500100
y: -19.996000 20.004000
z: -0.999800 1.000200
bbox centre @ 0.000100 0.004000 0.000200
diff of file:
=============</code>

Then librat has been installed correctly